### ISSUE TITLE
Record a task's parameters using a ``configuration`` section in the analysis file

### ISSUE THREAD

**Issue by dvandyk** [eos/eos#885]
In a step toward reproducible analyses, tasks should be able to resort to a pre-recorded configuration. This configuration should be stored inside the analysis file. To keep this as light-weight as possible, this should probably be done in a two-step process:

**matching**: a task matches the its name and parameters to the contents of the ``matching`` section of configuration. Matching must involve all mandatory parameters. Matching can involve optional parameters. A matched optional parameter must not appear in the ``setting`` section. More specific matching statements override less specific matching statements.

**setting**: a task overrides all non-mandatory parameters with the contents of the ``setting`` section of the configuration.

Three mock examples:

1. The task ``sample-nested``, when executed for the posteriors ``TH-all`` or ``CKM-all``, should set the argument ``dlogz`` to ``1.0`` and ``nlive`` to ``300``. However, when executed for the posterior ``WET-all``, the task should set the argument ``dlogz`` to ``1.0`` and ``nlive`` to ``500``. This example should be achived using the configuration:
```yaml
configuration:
  sample-nested:
    - match: { posterior: 'TH-all|CKM-all' }
      set: { dlogz: 1.0, nlive: 300 }
    - match: { posterior: 'WET-all' }
      set: { dlogz: 1.0, nlive: 500 }
```

2. The task ``find-mode`` expects a single mandatory argument (``posterior``). However, depending on the specific starting point (or method to determine the starting point), settings might require adjustment.
```yaml
configuration:
  find-mode:
    - match: { posterior: 'CKM-all', label: 'from-samples' }
      set: { optimizations: 25 }
    - match: { posterior: 'CKM-all' }
      set: { optimizations: 10 }
```

3. The task ``predict-observables`` expects two mandatory arguments (``posterior`` and ``prediction``). 
```yaml
configuration:
  predict-observables:
    - match: { posterior: 'SM-all|CKM-all|WET-all', prediction: 'BR-comparison' }
      set: { some_optional_parameter: 'foo' }
```

Looking forward to comments @mreboud @MJKirk @cbolognani @smeiser @martabm99

**Comment by lorenzennio** [comment]
I think that optional arguments of tasks should become mandatory here. This ensures backwards compatibility when defaults of optional arguments should change.

**Comment by dvandyk** [comment]
Add support for
```
eos-analysis list-posteriors -c / --configured-only
eos-analysis list-predictions -c / -- configured-only
```
to list only posteriors and predictions that have a configuration. Will be useful for the snakemake file.

**Comment by MJKirk** [comment]
Closed by #950
